tests/core/kernel-and-base-single-reboot/task.yaml

```diff
@@ -0,0 +1,132 @@
+        cp /bin/systemctl /tmp/orig-systemctl
+        mount -o bind "$TESTSLIB/mock-shutdown" /bin/systemctl
+        tests.cleanup defer umount /bin/systemctl
+
+        snap refresh --no-wait "$core_snap" pc-kernel > refresh-change-id
+        test -n "$(cat refresh-change-id)"
+        change_id="$(cat refresh-change-id)"
+        # wait until we observe reboots
+        # shellcheck disable=SC2016
+        retry -n 100 --wait 5 sh -c 'test "$(wc -l < /tmp/mock-shutdown.calls)" -gt "1"'
+        # stop snapd now to avoid snapd waiting for too long and deciding to
+        # error out assuming a rollback across reboot
+        systemctl stop snapd.service snapd.socket
+
+        # both link snaps should be done now, snapd was stopped, so we cannot
+        # use 'snap change' and we need to inspect the state directly (even if
+        # snapd was up, it would not respond to API requests as it would be busy
+        # retrying auto-connect)
+        snap debug state --change "$change_id" /var/lib/snapd/state.json > tasks.state
+        # both link snaps are done
+        MATCH ' Done\s+.*Make snap "pc-kernel" .* available' < tasks.state
+        MATCH " Done\s+.*Make snap \"$core_snap\" .* available" < tasks.state
+        # auto-connect of the base is in doing and waiting for reboot
+        MATCH " Doing\s+.*Automatically connect eligible plugs and slots of snap \"$core_snap\"" < tasks.state
+        # auto-connect of the kernel is still queued
+        MATCH ' Do\s+.*Automatically connect eligible plugs and slots of snap "pc-kernel"' < tasks.state
+
+        if os.query is-core18; then
+            snap debug boot-vars > boot-vars.dump
+            MATCH 'snap_mode=try' < boot-vars.dump
+            MATCH 'snap_try_core=core18_.*.snap' < boot-vars.dump
+            MATCH 'snap_try_kernel=pc-kernel_.*.snap' < boot-vars.dump
+        elif os.query is-core20; then
+            stat /boot/grub/kernel.efi | MATCH 'pc_kernel.*.snap/kernel.efi'
+            stat -L /boot/grub/kernel.efi
+            stat /boot/grub/try-kernel.efi | MATCH 'pc_kernel.*.snap/kernel.efi'
+            stat -L /boot/grub/try-kernel.efi
+        else
+            echo "unsupported Ubuntu Core system"
+            exit 1
+        fi
+
+        # restore shutdown so that spread can reboot the host
+        tests.cleanup pop
+
+        REBOOT
+    elif [ "$SPREAD_REBOOT" = 1 ]; then
+        change_id="$(cat refresh-change-id)"
+        # XXX: is this sufficiently robust?
+        snap watch "$change_id" || true
+        snap changes | MATCH "$change_id\s+(Done|Error)"
+        # we expect re-refresh to fail since the tests uses a fake store
+        snap change "$change_id" > tasks.done
+        MATCH '^Error .* Handling re-refresh' < tasks.done
+        # no other errors
+        grep -v 'Handling re-refresh' < tasks.done | NOMATCH '^Error'
+        # nothing was undone
+        grep -v 'Handling re-refresh' < tasks.done | NOMATCH '^Undone'
```

Comment:
Should we in addition have a sanity check that revisions of the snaps are as expected?

Reply:
Added